### Week 06
### Natural gas prices 
### https://datahub.io/core/natural-gas#resource-monthly
### Dataset contains Monthly prices of Natural gas, starting from January 1997 to current year. Prices are in nominal dollars. Time series of major Natural Gas Prices including US Henry Hub. Data comes from U.S. Energy Information Administration EIA Data Dataset contains Monthly prices of Natural gas, starting from January 1997 to current year. 

### In this dataset the  feature column is Month and the target column is Price

In [95]:
# Import necessary libraries
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.linear_model import LogisticRegression
from sklearn.svm import SVC
from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import classification_report, f1_score, accuracy_score, recall_score, precision_score, roc_auc_score, confusion_matrix, r2_score, ConfusionMatrixDisplay
from sklearn.utils import resample
from sklearn.model_selection import GridSearchCV
from sklearn.model_selection import RandomizedSearchCV
from sklearn.neighbors import KNeighborsRegressor
from sklearn.metrics import mean_squared_error



In [96]:
# Load the dataset
Dataset = pd.read_csv(r'rr.csv')

### From the above, I have taken Dataset as the variable and imported the data from the csv file through pd.read_csv command 

### Train and Test Sets

### A method for assessing a machine learning algorithm's performance is the train-test split. It may be applied to issues involving classification or regression as well as any supervised learning technique.The process entails splitting the dataset into two subgroups. The training dataset is the initial subset, which is used to fit the model. The model is not trained using the second subset; rather, it is given the input element of the dataset, and its predictions are then produced and contrasted with the expected values. The test dataset is the second dataset in question.
### Train data set: The machine learning model is fitted using the train dataset.
### Test data set: Used to assess how well a machine learning model fits the data.

In [98]:
# Split the resampled data into train and test sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=50)
X_train
X_test
y_train
y_test

122    0
82     0
97     0
99     1
67     1
21     0
174    0
163    0
169    0
210    0
25     0
137    1
255    1
179    0
103    1
237    0
128    0
109    0
282    1
86     1
250    1
170    0
223    0
35     0
89     0
246    0
176    0
266    1
9      1
108    0
40     1
256    0
26     1
211    0
64     1
232    0
267    0
123    1
32     0
19     0
16     0
209    1
203    1
90     0
136    1
34     0
17     1
74     0
120    0
245    0
202    0
152    0
217    0
195    1
38     1
125    0
83     1
Name: Price_inc, dtype: int32

### Applying Logistic Regression Model

### Logistic Regression: It is also referred as a predictive analytics and categorization frequently make use of this kind of statistical model, commonly referred to as a logit model. Based on a given dataset of independent variables, logistic regression calculates the likelihood of the event occurence. Given that the result is a probability, the dependent variable's range is 0 to 1. 

### Below, I have converted the data type of the month from object type to date time in order to perform logistic regression(as it will not allow object data type). As logistic regression works on the discrete type of data.. i have pulled up the price data as it is continuous from the csv file, and aimed to create a proce difference in a way it is compatible to perform logistic regression. 

In [99]:
# Logistic Regression

Dataset = pd.read_csv(r'C:/Users/vadla/Downloads/monthly_csv.csv')

df = pd.DataFrame(Dataset)

# Convert 'Month' to datetime and extract the month number
df['Month'] = pd.to_datetime(df['Month'])
df['Month_dt'] = df['Month'].dt.month

# Calculate the binary outcome based on price increase or decrease
df['Price_dif'] = df['Price'].diff()
df['Price_inc'] = (df['Price_dif'] > 0).astype(int)
df.dropna(inplace=True)  # drop the first row since it doesn't have a previous month to compare to

# Splitting data into training and testing set
X = df[['Month_dt']]
y = df['Price_inc']

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Fitting a Logistic regression model
logreg_model = LogisticRegression()
logreg_model.fit(X_train, y_train)
y_pred = logreg_model.predict(X_test)

accuracy_lr = accuracy_score(y_test, y_pred)
confusionmatrix_logreg = confusion_matrix(y_test, y_pred)
classificationreport_logreg = classification_report(y_test, y_pred)

print(f"Logistic Regression Accuracy: {accuracy_lr*100:.2f}%")
print(f'Confusion Matrix- Logistic Regression:\n{confusionmatrix_logreg}')
print(f'Classification Report- Logistic Regression:\n{classificationreport_logreg}')



Logistic Regression Accuracy: 49.12%
Confusion Matrix- Logistic Regression:
[[24 13]
 [16  4]]
Classification Report- Logistic Regression:
              precision    recall  f1-score   support

           0       0.60      0.65      0.62        37
           1       0.24      0.20      0.22        20

    accuracy                           0.49        57
   macro avg       0.42      0.42      0.42        57
weighted avg       0.47      0.49      0.48        57



### Logistic Regression Accuracy: 49.12%
### Values for Logistic Regeression Model:
### True Positive: 4
### False Negative: 16
### True Negative: 24
### False Positive: 13

### Applying Decision Tree Model

### Decision tree model is applied in order to classify and perfrom regression tasks. In this model it is based mainly on inquiring a sequence of questions are used by the decision tree to generate decisions.

### Here, I have initialized and trained the data for the application decision tree model and evaluated the model's performance with the scores of accuracy, confusion matrix and classification matrix.

In [100]:
# Initializing and training the Decision Tree model
clf_dectree = DecisionTreeClassifier(random_state=50)
clf_dectree.fit(X_train, y_train)

# Evaluating the model's performance
predictedmodel_values = clf_dectree.predict(X_test)

# Calculate accuracy
accuracy_dectree = accuracy_score(y_test, predictedmodel_values)

# Calculate confusion matrix
confusionmatrix_dectree = confusion_matrix(y_test, predictedmodel_values)

# Calculate classification report
classificationreport_dectree = classification_report(y_test, predictedmodel_values)

# Print the results
print(f'Decision Tree Model Accuracy: \n{accuracy_dectree*100:.2f}%')
print(f'Classification Report- decision tree model:\n{classificationreport_dectree}')
print(f'Confusion Matrix- decision tree model:\n{confusionmatrix_dectree}')


Decision Tree Model Accuracy: 
56.14%
Classification Report- decision tree model:
              precision    recall  f1-score   support

           0       0.64      0.76      0.69        37
           1       0.31      0.20      0.24        20

    accuracy                           0.56        57
   macro avg       0.47      0.48      0.47        57
weighted avg       0.52      0.56      0.53        57

Confusion Matrix- decision tree model:
[[28  9]
 [16  4]]


### Decision Tree Model Accuracy: 56.14%
### Values for Decision Tree Model:
### True Positive: 4 
### False Positive: 9
### True Negative: 28
### False Negative: 16

### Applying SVC Model

### The SVM can be defined as the Support Vector Machine, it is used to apply to classify tasks which can be defined as Support Vector Classification (SVC). THe comination of supervised learning techniques (SVM) will be applied in order to identify the regression, classification and outliers.

### Here, I have applied SVC with linear kernel and trained the model to predict the values in a way to calculate the scores of the accuracy, confusion matrix and classification report.

In [101]:
# Initialize the model with a linear kernel
clf_linear = SVC(kernel='linear')
# Train the model
clf_linear.fit(X_train, y_train)
# Make predictions
y_pred_linear = clf_linear.predict(X_test)

In [102]:
# Accuracy
accuracy_linear = accuracy_score(y_test, y_pred_linear)

# Confusion Matrix
confusionmatrix_linear = confusion_matrix(y_test, y_pred_linear)

# Classification Report
classificationreport_linear = classification_report(y_test, y_pred_linear)

print(f'Linear Kernel:(Classification Report)- \n{classificationreport_linear}')
print(f'Linear Kernel:(Accuracy)- \n{accuracy_linear*100:.2f}%')
print(f'Linear Kernel:(Confusion Matrix)- \n{confusionmatrix_linear}')

Linear Kernel:(Classification Report)- 
              precision    recall  f1-score   support

           0       0.65      1.00      0.79        37
           1       0.00      0.00      0.00        20

    accuracy                           0.65        57
   macro avg       0.32      0.50      0.39        57
weighted avg       0.42      0.65      0.51        57

Linear Kernel:(Accuracy)- 
64.91%
Linear Kernel:(Confusion Matrix)- 
[[37  0]
 [20  0]]


C:\Users\vadla\anaconda3\Lib\site-packages\sklearn\metrics\_classification.py:1469: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
C:\Users\vadla\anaconda3\Lib\site-packages\sklearn\metrics\_classification.py:1469: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
C:\Users\vadla\anaconda3\Lib\site-packages\sklearn\metrics\_classification.py:1469: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))


### Values for SVC Model:
### SVC Linear Kernel Accuracy: 64.91% 
### True Positive: 0
### False Negative: 20
### True Negative: 37
### False Positive: 0

### Inferences:
### To conclude, out of the three models, the accuracy value of SVC Model is 64.91% which is highest among the three models. It is preffered model to identify the patterns of changes in the natural gas prices on the monthly basis